### PR TITLE
testapp config tests and automated selenium testing

### PR DESCRIPTION
**Pull request by timlegge**

### CODE REVIEW

File: xt/testapp/t/003_test_idps_configuration.t [hunk 1]
```diff
@@ -0,0 +1,63 @@
+use strict;
```
Comment (waterkip) [inline xt/testapp/t/003_test_idps_configuration.t:1]: This block can be replaced with `use Test::Lib; use Test::Net::SAML2;`

File: xt/testapp/t/003_test_idps_configuration.t [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+use warnings;
+use Test::More;
+use Data::Dumper;
```
Comment (waterkip) [inline xt/testapp/t/003_test_idps_configuration.t:4]: Test::More has `explain`.

File: xt/testapp/t/004_selenium.t [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+Value of the issuer to find on the user logged in page for Net::SAML2 testapp to
+confirm that the login worked
+
+=back
+
+=cut
+
+my $driver =  Selenium::Chrome->new(
+                        driver => '~/perl-Net-SAML2/xt/testapp/t/selenium/chromedriver',
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:120]: This is very specific to your environment, you can use Find::Bin for this.

File: xt/testapp/t/004_selenium.t [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+##############################################
+# Loop through each of the IdPs being tested
+##############################################
+foreach my $idp (@idps) {
+    print "==============================================\n";
+    print "= Beginning test for IdP: $idp\n";
+    print "==============================================\n";
+
+    # Load the IdP configuration file
+    my $config_file = 'IdPs/' . $idp . '/selenium.yml';
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:145]: I would use `File::Spec::Functions qw(catfile)` for this. Also makes paths Windows compatible.

File: xt/testapp/t/004_selenium.t [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+            }
+            print "    Login Complete\n";
+
+            # Check the value of the issuer if issuer_value is defined in the config
+            my $value = $driver->find_element_by_id("issuer");
+            if ( defined $selenium_config->{"issuer_value"} ) {
+                ok ($selenium_config->{"issuer_value"} eq $value->get_text(), "Login confirmed: $selenium_config->{'issuer_value'} found");
+            }
+
+            print "    Logout using: $logout_binding", "\n";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:248]: As this is a testfile, I would opt to use `diag` or `note`.

File: xt/testapp/t/003_test_idps_configuration.t (hunk 1 above)
Comment (timlegge) [reply xt/testapp/t/003_test_idps_configuration.t:1]: perl -I ../../blib/lib/ -I blib/lib/ t/003_test_idps_configuration.t
unable to find t/lib directory in /home/tim/perl-Net-SAML2/xt/testapp/t/ at /home/tim/perl5/lib/perl5/Test/Lib.pm line 32.
BEGIN failed--compilation aborted at t/003_test_idps_configuration.t line 1.

It appears relative to the current directory.  So I would need to copy a bunch of code to xt/testapp/t

File: xt/testapp/t/003_test_idps_configuration.t (hunk 2 above)
Comment (timlegge) [reply xt/testapp/t/003_test_idps_configuration.t:4]: I will look at that

File: xt/testapp/t/004_selenium.t (hunk 4 above)
Comment (timlegge) [reply xt/testapp/t/004_selenium.t:145]: will do

File: xt/testapp/t/004_selenium.t (hunk 5 above)
Comment (timlegge) [reply xt/testapp/t/004_selenium.t:248]: sounds good

**Review by timlegge — COMMENTED** [review]
LGTM

File: xt/testapp/t/004_selenium.t [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+my $driver =  Selenium::Chrome->new(
+                        binary => "$RealBin/selenium/chromedriver",
+                        accept_ssl_certs  => 1,
+                        );
+
+isa_ok($driver, 'Selenium::Chrome');
+
+my $ret = $driver->get('https://netsaml2-testapp.local');
+if ( ! $ret ) {
+    die "Unable to access https://netsaml2-testapp.local";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:141]: I would perhaps use BAILOUT from Test::More here.

File: xt/testapp/t/004_selenium.t [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+=item * L<issuer_value>
+
+Value of the issuer to find on the user logged in page for Net::SAML2 testapp to
+confirm that the login worked
+
+=back
+
+=cut
+if (! -d "$RealBin/selenium" ) {
+    warn "Created directory $RealBin/selenium\n" if (make_path("$RealBin/selenium"));
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:124]: Is this a warn or more some kind of informational message?

File: xt/testapp/t/004_selenium.t [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+if (! -e -f "$RealBin/selenium/chromedriver") {
+    warn "Please ensure that the chromedriver binary is in $RealBin/selenium/\n";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:128]: Perhaps BAILOUT as well.

File: xt/testapp/t/004_selenium.t [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+$driver->quit;
+
+sub auth_to_idp {
+    my $idp = shift;
+
+    note "==============================================\n";
+    note "= Beginning test for IdP: $idp\n";
+    note "==============================================\n";
+    # Load the IdP configuration file
+    my $config_file = catfile( "IdPs/$idp/selenium.yml");
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:170]: catfile works with an array to construct the correct path: `catfile(qw(IdPS foo selenium.yml));`

File: xt/testapp/t/004_selenium.t [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+    if ( ! -e $config_file ) {
+        plan skip_all => "IdPs/$idp/selenium.yml does not exist";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:172]: `plan skip_all => $config_file does not exist`.

File: xt/testapp/t/004_selenium.t [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        next;
+    }
+    ok (defined $config_file, "selenium config file found for $idp");
+
+    my $selenium_config = YAML::LoadFile($config_file);
+    if ( ! $selenium_config ) {
+        die "Undefined selenium_config";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:179]: BAILOUT? Or `return fail("Undefined selenium config");`

File: xt/testapp/t/004_selenium.t [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+    }
+    ok ($selenium_config, "Loaded selenium config from $config_file");
+
+    ############################################
+    # Loop through each of the logout bindings
+    ############################################
+    foreach my $logout_binding (@{$selenium_config->{"logout_bindings"}}) {
+        note "Logout: ", $logout_binding, "\n";
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:187]: I believe `note` is always showing when running the testsuite and `diag` is only when running with `-v`.

File: xt/testapp/t/004_selenium.t [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+};
+
+sub load_idps {
+    if ( ! -x './IdPs' ) {
+        return "<html><pre>You must have a xt/testapp/IdPs directory</pre></html>";
+    }
+    my @dirs = read_dir('./IdPs');
+    my @idps;
+    for my $dir (sort @dirs) {
+        if ( $dir eq '.keep' ) { next ; }
```
Comment (waterkip) [inline xt/testapp/t/004_selenium.t:306]: `if -d $dir` would perhaps be easier?

File: xt/testapp/t/004_selenium.t (hunk 7 above)
Comment (timlegge) [reply xt/testapp/t/004_selenium.t:124]: informational - could the  print

File: xt/testapp/t/004_selenium.t (hunk 11 above)
Comment (timlegge) [reply xt/testapp/t/004_selenium.t:179]: skip_all here will skip all the tests in the subtest.  The tests did not fail the config was just wrong

**Pull request merged**